core/drivers/crypto/hisilicon/hpre_main.c

```diff
@@ -0,0 +1,319 @@
+	io_write32(qm->io_base + HPRE_BD_ARUSR_CFG, HPRE_BD_ARUSR_MASK);
+	io_write32(qm->io_base + HPRE_BD_AWUSR_CFG, HPRE_BD_AWUSR_MASK);
+	io_write32(qm->io_base + HPRE_DATA_RUSER_CFG, HPRE_DATA_USR_MASK);
+	io_write32(qm->io_base + HPRE_DATA_WUSER_CFG, HPRE_DATA_USR_MASK);
+	io_write32(qm->io_base + HPRE_TYPES_ENB, HPRE_RSA_ENB | HPRE_ECC_ENB);
+	io_write32(qm->io_base + HPRE_RDCHN_INI_CFG, 0x1);
+	ret = IO_READ32_POLL_TIMEOUT(qm->io_base + HPRE_RDCHN_INI_ST, val,
+				     val & 0x1, POLL_PERIOD, POLL_TIMEOUT);
+	if (ret) {
+		EMSG("Fail to init rd channel");
+		return TEE_ERROR_BUSY;
+	}
+
+	ret = hpre_set_cluster(qm);
+	if (ret) {
+		EMSG("Fail to init hpre cluster cores");
+		return TEE_ERROR_BUSY;
+	}
+
+	hpre_enable_clock_gate(qm);
+
+	return TEE_SUCCESS;
+}
+
+static enum hisi_drv_status hpre_dev_status_check(struct hisi_qm *qm)
+{
+	uint32_t val = 0;
+
+	val = io_read32(qm->io_base + HPRE_HAC_INT_SRC);
+	if (val & HPRE_RAS_NFE_MASK) {
+		EMSG("HPRE NFE RAS happened, need to reset");
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	val = io_read32(qm->io_base + HISI_QM_ABNML_INT_SRC);
+	if (val) {
+		if (val & HISI_QM_HPRE_NFE_INT_MASK)
+			EMSG("QM NFE RAS happened, need to reset");
+
+		if (val & HISI_QM_INVALID_DB) {
+			EMSG("QM invalid db happened, please check");
+			io_write32(qm->io_base + HISI_QM_ABNML_INT_SRC,
+				   HISI_QM_INVALID_DB);
+		}
+
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status hpre_qm_init(struct acc_device *hpre_dev)
+{
+	struct hisi_qm *qm = &hpre_dev->qm;
+
+	if (cpu_mmu_enabled()) {
+		qm->io_base = (uintptr_t)phys_to_virt_io(hpre_dev->io_base,
+							 hpre_dev->io_size);
+		if (!qm->io_base) {
+			EMSG("Fail to get qm io_base");
+			return HISI_QM_DRVCRYPT_EFAULT;
+		}
+	} else {
+		qm->io_base = hpre_dev->io_base;
+	}
+
+	qm->vfs_num = hpre_dev->vfs_num;
+	qm->fun_type = hpre_dev->fun_type;
+	qm->sqe_size = HPRE_SQE_SIZE;
+	qm->sqe_log2_size = HPRE_SQE_LOG2_SIZE;
+	if (qm->fun_type == HISI_QM_HW_PF) {
+		hisi_qm_get_version(qm);
+		DMSG("HPRE hardware version is 0x%"PRIx32, qm->version);
+		qm->qp_base = HISI_QM_PF_Q_BASE;
+		qm->qp_num = HISI_QM_PF_Q_NUM;
+		qm->dev_status_check = hpre_dev_status_check;
+	}
+
+	return hisi_qm_init(qm);
+}
+
+static struct acc_device *hpre_pre_init(void)
+{
+	struct acc_device *hpre_dev = NULL;
+
+	hpre_dev = calloc(1, sizeof(*hpre_dev));
+	if (!hpre_dev) {
+		EMSG("Fail to alloc hpre_dev");
+		return NULL;
+	}
+
+	hpre_dev->io_base = HPRE_BAR_BASE;
+	hpre_dev->io_size = HPRE_BAR_SIZE;
+	hpre_dev->vfs_num = 0;
+	hpre_dev->endian = 0;
+	hpre_dev->fun_type = HISI_QM_HW_PF;
+	SLIST_INSERT_HEAD(&hpre_list, hpre_dev, link);
+
+	return hpre_dev;
+}
+
+static TEE_Result hpre_probe(void)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	struct acc_device *hpre_dev = NULL;
+	struct hisi_qm *qm = NULL;
+
+	DMSG("HPRE driver init start");
+	hpre_dev = hpre_pre_init();
+	if (!hpre_dev)
+		return TEE_ERROR_STORAGE_NOT_AVAILABLE;
```

Comment:
Is this really about TEE storage? Seen the error case in `hpre_pre_init()`, prefer `TEE_ERROR_OUT_OF_MEMORY`.